### PR TITLE
Adds labels support for install and upgrade commands

### LINKED ISSUE #7007: Add custom labels to helmv3 release secret
Hello,

Is there a way to add a specific label to the helm release secret created by Helmv3 ? I read all the docs and browsed issued without success.

My use case is really simple, I use Velero to backup all my stuff, project by project (1 project = 1 helm deployment), every day according specific labels. If I can add specific labels, I can backup helm information at the same time.

I can do a workaround with a hook I know but I thing that feature could be useful for others.

### PR DESCRIPTION
**Pull request by dm3ch**
/closes https://github.com/helm/helm/issues/7007

**What this PR does / why we need it**:
Adding support of providing custom labels during install/update of release that could be used for selector filtering introduced in https://github.com/helm/helm/pull/8532.

**Special notes for your reviewer**:
This PR focused on adding custom label support during install/upgrade actions.
There was similar PRs before that, unfortunately, haven't been finished: https://github.com/helm/helm/pull/8546 https://github.com/helm/helm/pull/8099 https://github.com/helm/helm/pull/9087

In this PR I'm trying to make changes as small as possible, to make development and review easier.
My goal is to make selector functionality in the list command useable for install/upgrade actions with secret and configmap storage drivers.

I don't want to add SQL support in this particular PR, cause it seems it would at least double the number of changes.

TODO:
Functionality (support labels argument during following actions):
- [x] install
- [x] upgrade --install
- [x] upgrade (basic)
- [x] upgrade (keep labels on superseded releases)
- [x] tests for new functionality (don't know how to implement it better)
- [x] Secret storage driver
- [x] Configmap storage driver

Not planned in this PR:
- [x] SQL - support is not currently planned in this PR. But there are some related developments in https://github.com/helm/helm/pull/8546 that could be prepared as separate PR later.
- [x] Merge original release labels with upgrade labels

**UPD**:
Now I have implemented even features I haven't planned for this PR. So right now this PR is completely ready for review.

If you're interested why I've created a separate DB table (in SQL driver) for custom releases - my idea was that some day we may want to make release filtering on storage backend side (k8s and SQL support requests with filtering by labels/joined table). so I decided that it would be better to split it to a separate table instead of having it encoded somewhere in json, because data migration is always annoying.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by dm3ch** [comment]
Actually, I'm searching for help with following errors:
https://app.circleci.com/pipelines/github/helm/helm/3486/workflows/e06a9fd9-c72c-47a4-adfe-c8fca93560b2/jobs/15065

The problem is that I'm inserting multiple rows in `for r := range map[string]string` and I'm getting different orders for maps with same data in insert and test code places. Unfortunately, I haven't yet found a workaround for that, so this test flack sometime

**Comment by dm3ch** [comment]
@hickeyma what should I do to start the review process?

Sorry for the direct mention, but you are the person who helped me with making https://github.com/helm/helm/pull/8532 merged.

**Comment by hickeyma** [comment]
@dm3ch I have added to "awaiting review" label to it to notify maintainers that it needs a review. They should get to it when they have the bandwidth.

**Comment by dm3ch** [comment]
> @dm3ch I have added to "awaiting review" label to it to notify maintainers that it needs a review. They should get to it when they have the bandwidth.

Thank you

**Comment by koulq** [comment]
Hi, which version will this feature be put into

**Comment by dm3ch** [comment]
> Hi, which version will this feature be put into
Maintainers still haven't started review process, so I don't have even assumptions when this feature would be released

**Comment by dm3ch** [comment]
@hickeyma Sorry for the mention, but you are the only person who helped me previous time.

Is it normal that PR is hanging for such a long time without any reaction or I have done something wrong and need to ping core team to draw attention?

**Comment by hickeyma** [comment]
@dm3ch Do you mind doing a rebase to resolve the conflict?

**Comment by dm3ch** [comment]
> @dm3ch Do you mind doing a rebase to resolve the conflict?

Done, but as far as I remember there was one flaky test (some of SQL one), that I wasn't able to stabilize myself. So it's possible that now test CI job can fail.

**Comment by dm3ch** [comment]
And it failed on that particular flaky test with SQL. :)

**Comment by dm3ch** [comment]
I finally found to prevent SQL storage driver test from flaking.

**Comment by dm3ch** [comment]
@hickeyma Sorry again for mentioning you. Do you have any other recomendations?

**Comment by izturn** [comment]
any progress ?

**Comment by dm3ch** [comment]
> any progress?

I think this PR is ready. But I'm having problems with getting a review.
So I believe there's no progress. :(

**Comment by izturn** [comment]
> 

:(, thx

**Comment by izturn** [comment]
> > any progress?
> 
> I think this PR is ready. But I'm having problems with getting a review. So I believe there's no progress. :(

:(, thx

**Comment by wjur** [comment]
@hickeyma any chances that this PR gets review soon?

**Comment by joyme123** [comment]
this is a useful feature

File: cmd/helm/install.go
Comment (huguesalary) [inline cmd/helm/install.go:158]:
```suggestion
	f.StringToStringVarP(&client.Labels, "labels", "l", nil, "Labels that would be added to release metadata. Should be divided by comma.")
```

File: cmd/helm/upgrade.go
Comment (huguesalary) [inline cmd/helm/upgrade.go:232]:
```suggestion
	f.StringToStringVarP(&client.Labels, "labels", "l", nil, "Labels that would be added to release metadata. Should be separated by comma. Original release labels will be merged with upgrade labels. You can unset label using null.")
```

**Comment by dm3ch** [comment]
@huguesalary Sorry, I missed your suggestions during my Christmas holidays.
Just applied your suggestions and merged the actual main branch to fix conflicts.

**Comment by huguesalary** [comment]
> @huguesalary Sorry, I missed your suggestions during my Christmas holidays. Just applied your suggestions and merged the actual main branch to fix conflicts.

LGTM, though I am not an official maintainer :) so you'll need to get a maintainer's review to merge this

**Comment by dm3ch** [comment]
@mattfarina @joejulian Hi! I see that you've added milestone labels to this PR. Could you please help us find a maintainer willing to review this PR and who has some free time for that?

**Comment by joejulian** [comment]
I'm doing my best. :/

**Comment by tiancandevloper** [comment]
The feature of adding labels is also required during a rollback, so it is best to add the label of currentRelease when constructing the targetRelease in the rollback process.

**Comment by joejulian** [comment]
I tried using this feature and setting one of the built-in labels. It didn't overwrite them, but it didn't error either. I would have expected an error if I try to use a reserved label.

**Comment by dm3ch** [comment]
Makes sense, I would try to add this check this week

**Review by joejulian — APPROVED** [review]
LGTM. Needs one more approval.

File: pkg/action/install.go
Comment (sabre1041) [inline pkg/action/install.go:260]: Would we be able to remove this debugging line?

**Review by sabre1041 — CHANGES_REQUESTED** [review]
Looks good. One small change and we should be good

File: pkg/action/install.go
Comment (dm3ch) [reply pkg/action/install.go:260]:
```suggestion
	fmt.Println(driver.ContainsSystemLabels(i.Labels))
```

Comment (dm3ch) [reply pkg/action/install.go:260]:
```suggestion
	fmt.Println(driver.ContainsSystemLabels(i.Labels))
```

Comment (dm3ch) [reply pkg/action/install.go]: Completely forgot about it, removed it. TY

**Review by sabre1041 — APPROVED** [review]
LGTM

**Comment by dm3ch** [comment]
So what is the our next steps now when we got 2 approves on this PR? 🤔

**Comment by yulin-li** [comment]
Waiting for this useful feature. Could this be merged?

**Comment by dm3ch** [comment]
@joejulian @sabre1041 Now this PR have 2 approves, what is the next steps?

**Comment by joejulian** [comment]
I'm just going to give it one more CI run after merging in main. Normally, the second approver would merge.

**Comment by dm3ch** [comment]
So, CI run went fine. Am I right understood that @sabre1041 was supposed to merge this PR? :)

**Comment by dm3ch** [comment]
🎉 
@sabre1041 @joejulian Thank you a lot for review and merge.

File: pkg/storage/driver/secrets.go
Comment (hiddeco) [inline pkg/storage/driver/secrets.go:102]: This change appears to have caused a breakage of the `helm list --selector` feature, see: https://github.com/helm/helm/issues/12435

Comment (dm3ch) [reply pkg/storage/driver/secrets.go:102]: So desired behaviour is allowing users to filter using system labels?

Comment (hiddeco) [reply pkg/storage/driver/secrets.go:102]: Not a maintainer, but I imagine this can indeed by useful to select a subset of things using the default labels Helm provides you with.

It also appears to be what's being tested, which continued to work because the in-memory driver was not modified in this PR.

Comment (dm3ch) [reply pkg/storage/driver/secrets.go:102]: Let's continue this discussion in fix PR - https://github.com/helm/helm/pull/12447

**Pull request merged**
